0006-ffmul-revised.md

```diff
@@ -0,0 +1,518 @@
+
+$$\tag{$2'''$}
+a b - q f - r = 2^{3\ell} (w + c_1)
+$$
+
+
+
+By our first constraint $\text{(C1)}$, the LHS of $(2''')$ is a multiple of $n$, so we have
+
+
+$$
+2^{3\ell} (w + v_{1})  = 0 \mod{n}
+$$
+
+Because $2^{3\ell}$ and $n$ are co-prime, we can multiply with $2^{-3\ell}$ to see that
+
+$$
+w + v_{1}  = 0 \mod{n}
+$$
+
+Therefore, we can write $w + v_{1} = \delta n$ for some $\delta \in \mathbb{Z}$. From $(2''')$ we conclude that
+
+$$\tag{$2''''$}
+a b - q f - r = 2^{3\ell} n \delta
+$$
+
+#### Final bound
+
+There is one step remaining to make our gate sound: We must prove that $a b - q f - r$ cannot overflow $2^{3\ell} n$, so $\delta$ is in fact zero.
+
+To do this, our strategy is to add an additional bound on the highest limbs of $a$, $b$, $q$. Skipping the lower limbs and $r$ is fine because they all contribute a negligible amount to the overall estimate.
+
+For a variable $x = (x_0, x_1, x_2)$ define the _bounds check_ as the following two constraints:
+
+* $x_2 + (2^\ell - f_2 - 1) = z \mod n$
+* $z \in [0,2^\ell)$
+
+We write this succinctly as $x_2 + (2^\ell - f_2 - 1) \in [0,2^\ell)$ but we must not forget that "+" is finite field addition.
+
+For a field element $x_2 \in [0, n)$ the bounds check implies that
+
+$$x_2 \in [0, f_2 + 1) \cup [n - 2^\ell + f_2 + 1, n).$$
+
+If, in addition, $x_2$ is range-checked to be at most $\ell$ bits, the high interval is excluded:
+
+$$x_2 \in [0, 2^\ell) \quad\text{and}\quad x_2 + (2^\ell - f_2 - 1) \in [0,2^\ell) \quad\Longrightarrow\quad x_2 \in [0, f_2 + 1)$$
+
+> NB: It's important to use $x_2 + (2^\ell - f_2 - 1)$ and not $x_2 + (2^\ell - f_2)$ in the bounds check, so that $x_2$ can be at most $f_2$ and not $f_2 - 1$. Otherwise, the check would exclude valid foreign field elements for which $x_2 = f_2$.
+
+For $x = (x_0, x_1, x_2)$, an $\ell$-bit range check on all limbs plus bounds check imply $x < 2^{2\ell}(f_2 + 1)$, as follows:
+
+$$
+x = 2^{2\ell} x_2 + 2^\ell x_1 + x_0  < 2^{2\ell} f_2 + 2^{2\ell} = 2^{2\ell}(f_2 + 1)
+$$
+
+This estimate is what we need for $a$, $b$ and $q$. We assume that bounds checks happen externally for $a$ and $b$:
+
+$$
+\begin{align}
+\tag{Bounds check: $a$}
+& a_2 + (2^\ell - f_2 - 1) \in [0,2^\ell) \\
+\tag{Bounds check: $b$}
+& b_2 + (2^\ell - f_2 - 1) \in [0,2^\ell) \\
+\end{align}
+$$
+
+To help with the bounds check on $q$, we use some free space in the ffmul gate to expose $q'_2 := q_2 + (2^\ell - f_2 - 1)$ as a witness.
+
+* **Introduce 1 witness**: $q'_2$
+
+The addition is another constraint:
+
+$$\tag{C11: bound for $q$}
+q'_2 = q_2 + (2^\ell - f_2 - 1) \mod n
+$$
+
+And a range check on $q'_2$ completes the bounds check on $q$:
+
+$$\tag{RC: bound for $q$}
+q'_2 \in [0,2^\ell)
+$$
+
+This implies that $a,b,q \in[0, 2^{2\ell}(f_2 + 1))$. For $r$, we already know that $r \in [0, 2^{3\ell})$.
+
+We now get the following upper bound and lower bound:
+
+$$
+ab - qf - r \le ab < 2^{4\ell}(f_2 + 1)^2
```

Comment:
Perhaps let's add a middle `ab ≤ (2^{2\ell} * f_2 + 2^\ell * 2^\ell + 2^\ell )^2`, then that gives you a more accurate upper bound that `ab < 2^{4\ell}(f_2+1)^2 + 2^2\ell + 2*2^\ell*2^\ell*(f_2 + 1)`

Reply:
but $ab < 2^{4\ell}(f_2 + 1)^2$ is correct, because $a, b < 2^{2\ell}(f_2 + 1)$ as I showed above. your bound is less tight

Reply:
note that $2^\ell a_1 + a_0 \le 2^\ell (2^\ell - 1) + (2^\ell - 1) = 2^{2\ell} - 1$, so estimating that part with $< 2^{2\ell}$ is perfect